aggregator.go

```diff
@@ -0,0 +1,300 @@
+			return cid.Undef, err
+		}
+		d.PathIndexes[2] = len(innerNodes[parentName].Links()) - 1
+	}
+
+	newBlocks := make([]ipldformat.Node, 0, 1<<20)
+
+	// secondary layer, 2 bytes off end ( drop the 2 second-to-last )
+	outerNodes := make(nodeMap)
+	for _, nodeName := range sortedNodeNames(innerNodes) {
+		nd := innerNodes[nodeName]
+
+		newBlocks = append(newBlocks, nd)
+		parentName := nodeName[0:len(nodeName)-4] + nodeName[len(nodeName)-2:]
+		if _, exists := outerNodes[parentName]; !exists {
+			outerNodes[parentName] = emptyDir()
+		}
+
+		if err := outerNodes[parentName].AddNodeLink(nodeName, nd); err != nil {
+			return cid.Undef, err
+		}
+
+		for _, innerDaglink := range nd.Links() {
+			d := dags[innerDaglink.Name]
+			d.PathPrefixes[1] = nodeName
+			d.PathIndexes[1] = len(outerNodes[parentName].Links()) - 1
+		}
+	}
+
+	// root
+	root := emptyDir()
+	for _, nodeName := range sortedNodeNames(outerNodes) {
+
+		nd := outerNodes[nodeName]
+		newBlocks = append(newBlocks, nd)
+		if err := root.AddNodeLink(nodeName, nd); err != nil {
+			return cid.Undef, err
+		}
+
+		for _, outerDagLink := range nd.Links() {
+			for _, innerDaglink := range innerNodes[outerDagLink.Name].Links() {
+				d := dags[innerDaglink.Name]
+				d.PathPrefixes[0] = nodeName
+				d.PathIndexes[0] = len(root.Links()) // no ...-1, idx#0 is left for the manifest
+			}
+		}
+	}
+
+	// now that we have all the paths correctly, assemble the manifest and add it to the root
+	prdr, pwrr := io.Pipe()
+
+	errCh := make(chan error, 1)
+	go func() {
+
+		defer func() {
+			err := pwrr.Close()
+			if err != nil {
+				errCh <- err
+			}
+		}()
+
+		j := json.NewEncoder(pwrr)
+
+		err := j.Encode(CurrentManifestPreamble)
+		if err != nil {
+			errCh <- err
+			return
+		}
+
+		err = j.Encode(ManifestSummary{
+			RecordType:      DagAggregateSummary,
+			EntriesSortedBy: "DagCidV1",
+			Description:     "Aggregate of non-related DAGs, produced by github.com/filecoin-project/go-dagaggregator-unixfs",
+			EntryCount:      len(sortedDags),
+		})
+		if err != nil {
+			errCh <- err
+			return
+		}
+
+		for _, d := range sortedDags {
+			err = j.Encode(d)
+			if err != nil {
+				errCh <- err
+				return
+			}
+		}
+	}()
+
+	leaves, err := (&importhelper.DagBuilderParams{
+		Dagserv:   ds,
+		RawLeaves: true,
+		Maxlinks:  8192 / 47, // importhelper.DefaultLinksPerBlock
+		CidBuilder: cid.V1Builder{
+			Codec:    uint64(multicodec.DagPb),
+			MhType:   uint64(multicodec.Sha2_256),
+			MhLength: 32,
+		},
+	}).New(chunker.NewSizeSplitter(prdr, 256<<10))
+	if err != nil {
+		return cid.Undef, err
+	}
+
+	manifest, err := balanced.Layout(leaves)
+	if err != nil {
+		return cid.Undef, err
+	}
+	if len(errCh) > 0 {
+		return cid.Undef, <-errCh
+	}
+
+	if err = root.AddNodeLink(AggregateManifestFilename, manifest); err != nil {
+		return cid.Undef, err
+	}
+
+	// we are done now, add everything
+	newBlocks = append(newBlocks, root)
+	err = ds.AddMany(context.Background(), newBlocks)
```

Comment:
Q: worth accepting a `ctx` in the method params?

Reply:
Hrm... yeah... I suppose...
We are not making many extra blocks, but... might as well...

Reply:
Done